File: modules/image/tests/test_image.cpp
/*
 * Copyright 2023 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
////////////////////////////////////////////////////////////////////////////////
#include <lagrange/image/RawInputImage.h>
#include <lagrange/image/image_type_conversion.h>

#include <lagrange/Logger.h>

#include <lagrange/testing/common.h>
////////////////////////////////////////////////////////////////////////////////

TEST_CASE("Image, boilerplate unit test", "[image]")
{
    // Very simple unit tests, basic checks

    std::vector<float> zero_image;
    zero_image.resize(1024, 0);

    LA_CHECK_THROWS(lagrange::image::make_default_rgba_image(16, 16, nullptr));
    LA_CHECK_THROWS(lagrange::image::make_default_rgba_image(0, 16, zero_image.data()));
    LA_CHECK_THROWS(lagrange::image::make_default_rgba_image(16, 0, zero_image.data()));


    lagrange::image::RawInputImage test2 =
        lagrange::image::make_default_rgba_image(16, 16, zero_image.data());
    lagrange::logger().info(
        "image test: {} x {}, {} rowbytes",
        test2.get_width(),
        test2.get_height(),
        test2.get_row_byte_stride());

    CHECK(test2.get_row_byte_stride() == 256);
}

TEST_CASE("Half precision view", "[image]")
{
    std::vector<float> zero_image;
    zero_image.resize(1024, 0);
    auto raw_img = lagrange::image::make_default_rgba_image(16, 16, zero_image.data());
    auto storage = lagrange::image::image_storage_from_raw_input_image(raw_img);

    auto float_view = std::make_shared<lagrange::image::ImageView<float>>(
        storage,
        raw_img.get_width(),
        raw_img.get_height());

    using Pixel = Eigen::Matrix<Eigen::half, 3, 1>;

    auto float16_view = std::make_shared<lagrange::image::ImageView<Pixel>>();

    float16_view->convert_from(*float_view, 1);

    REQUIRE(float16_view->get_precision() == lagrange::image::ImagePrecision::float16);
    REQUIRE(float16_view->get_channel() == lagrange::image::ImageChannel::three);
    REQUIRE(float16_view->get_view_stride_in_byte()(1) == sizeof(Pixel) * raw_img.get_width());

    lagrange::logger().info(
        "float16 image test: {} x {}, {} rowbytes",
        float16_view->get_view_size()(0),
        float16_view->get_view_size()(1),
        float16_view->get_view_stride_in_byte()(1));
}

File: modules/image/python/tests/test_image.py
#
# Copyright 2026 Adobe. All rights reserved.
# This file is licensed to you under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License. You may obtain a copy
# of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed under
# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
# OF ANY KIND, either express or implied. See the License for the specific language
# governing permissions and limitations under the License.
#
import lagrange
import pytest

import numpy as np


class TestImage:
    def test_split_grid_auto(self):
        # Grid shape is (H, W, C) = (256, 384, 3): 2 rows x 3 cols of 128x128 cells.
        grid = np.zeros((256, 384, 3), dtype=np.float32)
        for row in range(2):
            for col in range(3):
                grid[
                    row * 128 : (row + 1) * 128,
                    col * 128 : (col + 1) * 128,
                    :,
                ] = float(row * 3 + col + 1)

        cells = lagrange.image.split_grid(grid, 6)
        assert len(cells) == 6
        for idx, cell in enumerate(cells):
            assert cell.shape == (128, 128, 3)
            assert np.all(cell == float(idx + 1))

    def test_split_grid_explicit_layout(self):
        # Force a 1x6 layout: 6 cells of (128, 64, 3).
        grid = np.zeros((128, 384, 3), dtype=np.float32)
        for col in range(6):
            grid[:, col * 64 : (col + 1) * 64, :] = float(col + 1)

        cells = lagrange.image.split_grid(grid, 6, rows=1, cols=6)
        assert len(cells) == 6
        for idx, cell in enumerate(cells):
            assert cell.shape == (128, 64, 3)
            assert np.all(cell == float(idx + 1))

    def test_split_grid_partial_layout(self):
        # Specify only rows: cols is derived as num_cells / rows.
        grid = np.zeros((256, 256, 3), dtype=np.float32)
        for row in range(2):
            for col in range(2):
                grid[
                    row * 128 : (row + 1) * 128,
                    col * 128 : (col + 1) * 128,
                    :,
                ] = float(row * 2 + col + 1)

        cells = lagrange.image.split_grid(grid, 4, rows=2)
        assert len(cells) == 4
        for idx, cell in enumerate(cells):
            assert cell.shape == (128, 128, 3)
            assert np.all(cell == float(idx + 1))

    def test_split_grid_view_writes_through(self):
        # Mutating a returned cell view should mutate the source grid.
        grid = np.zeros((128, 256, 3), dtype=np.float32)
        cells = lagrange.image.split_grid(grid, 2)
        cells[1][:] = 7.0
        assert np.all(grid[:, 128:, :] == 7.0)
        assert np.all(grid[:, :128, :] == 0.0)

    def test_split_grid_zero_cells_raises(self):
        grid = np.zeros((128, 128, 3), dtype=np.float32)
        with pytest.raises(RuntimeError):
            lagrange.image.split_grid(grid, 0)

    def test_split_grid_indivisible_raises(self):
        # 100x100 grid cannot be evenly split into 3 cells.
        grid = np.zeros((100, 100, 3), dtype=np.float32)
        with pytest.raises(RuntimeError):
            lagrange.image.split_grid(grid, 3)

    def test_split_grid_explicit_mismatch_raises(self):
        # rows=2, cols=2 gives 4 cells but num_cells=3.
        grid = np.zeros((128, 128, 3), dtype=np.float32)
        with pytest.raises(RuntimeError):
            lagrange.image.split_grid(grid, 3, rows=2, cols=2)
